### ISSUE TITLE
Scale-down event of EKS kube-apiserver causes network outage

### ISSUE THREAD

**Issue by mhulscher** [cilium/cilium#20915]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### What happened?

AWS EKS scales its kube-apiserver(s) out and in. I notice that whenever a kube-apiserver is removed, the cilium-agent and -operator continue to make API requests to this now decommissioned kube-apiserver. What's annoying is that this kube-apiserver does not seem to close existing connections. Instead, it now returns a 401 Unauthorized as a response to all requests. This could also be considered an EKS bug. The result is that cilium can no longer create endpoints, allocate ENIs or IPs and that the CiliumLocalRedirectPolicy and DNS proxy seem to stop functioning; potentially causing a cluster-wide network outage until cilium connects to a working kube-apiserver

I know that it is caused by a kube-apiserver being removed because I am monitoring the endpoints in `kubectl -n default get endpoints kubernetes`. Whenever an endpoint is removed, the problems start.

Restarting the agent and the operator restores functionality. Without a restart it takes about 5 minutes before cilium connects to a working endpoint.

From the operator:
```
cilium-operator-78d4dc7dbb-lv8q2 cilium-operator error retrieving resource lock kube-system/cilium-operator-resource-lock: Unauthorized
cilium-operator-78d4dc7dbb-lv8q2 cilium-operator level=error msg="error retrieving resource lock kube-system/cilium-operator-resource-lock: Unauthorized" subsys=klo
```

I don't know exactly how to reproduce this. What I do know is that new EKS clusters see the most scaling events of the kube-apiserver. I reproduce it myself by creating a new cluster and then scheduling a lot of pods. This often leads to scaling events.

Our cilium is installed with the following helm values:

```
affinity:
  nodeAffinity:
    requiredDuringSchedulingIgnoredDuringExecution:
      nodeSelectorTerms:
      - matchExpressions:
        - key: kubernetes.io/os
          operator: In
          values:
          - linux
        - key: kubernetes.io/arch
          operator: In
          values:
          - amd64
        - key: eks.amazonaws.com/compute-type
          operator: NotIn
          values:
          - fargate
bpf:
  hostRouting: false
  masquerade: true
  tproxy: true
cni:
  chainingMode: none
  configMap: cni-configuration
  customConf: true
devices: eth0
enableCiliumEndpointSlice: true
endpointHealthChecking:
  enabled: false
eni:
  awsEnablePrefixDelegation: true
  ec2APIEndpoint: ec2.eu-west-1.amazonaws.com
  enabled: true
  instanceTagsFilter: aws:eks:cluster-name=ci-eks-4qrg9g
  updateEC2AdapterLimitViaAPI: true
healthChecking: false
hubble:
  enabled: true
  eventBufferCapacity: "8191"
  metrics:
    enabled: null
    serviceMonitor:
      enabled: true
  relay:
    affinity:
      podAntiAffinity:
        requiredDuringSchedulingIgnoredDuringExecution:
        - labelSelector:
            matchExpressions:
            - key: k8s-app
              operator: In
              values:
              - hubble-relay
          topologyKey: kubernetes.io/hostname
    enabled: true
    podDisruptionBudget:
      enabled: true
      maxUnavailable: 1
    replicas: 3
    rollOutPods: true
  ui:
    affinity:
      podAntiAffinity:
        requiredDuringSchedulingIgnoredDuringExecution:
        - labelSelector:
            matchExpressions:
            - key: k8s-app
              operator: In
              values:
              - hubble-ui
          topologyKey: kubernetes.io/hostname
    enabled: true
    ingress:
      annotations: {}
      enabled: true
      hosts:
      - hubble.example.com
    podDisruptionBudget:
      enabled: true
      maxUnavailable: 1
    replicas: 3
    rollOutPods: true
ipam:
  mode: eni
k8sServiceHost: C0F975EE951307D359FFC679CA1FDD1F.sk1.eu-west-1.eks.amazonaws.com
k8sServicePort: 443
kubeProxyReplacement: strict
l7Proxy: true
loadBalancer:
  serviceTopology: true
localRedirectPolicy: true
operator:
  affinity:
    podAntiAffinity:
      requiredDuringSchedulingIgnoredDuringExecution:
      - labelSelector:
          matchExpressions:
          - key: io.cilium/app
            operator: In
            values:
            - operator
          - key: name
            operator: In
            values:
            - cilium-operator
        topologyKey: kubernetes.io/hostname
  extraArgs:
  - --unmanaged-pod-watcher-interval=0
  podDisruptionBudget:
    enabled: true
    maxUnavailable: 1
  priorityClassName: system-cluster-critical
  prometheus:
    enabled: true
    serviceMonitor:
      enabled: true
  replicas: 3
  rollOutPods: true
  tolerations:
  - key: node.cilium.io/agent-not-ready
    operator: Exists
  - key: node.kubernetes.io/not-ready
    operator: Exists
priorityClassName: system-node-critical
prometheus:
  enabled: true
  serviceMonitor:
    enabled: true
svcSourceRangeCheck: "false"
tunnel: disabled
updateStrategy:
  type: OnDelete
```

### Cilium Version

```
❯ cilium version
cilium-cli: 0.12.1 compiled with go1.18.5 on linux/amd64
cilium image (default): v1.12.0
cilium image (stable): v1.12.0
cilium image (running): v1.12.0
```

### Kernel Version

Several:
```
5.15.54-25.126.amzn2.x86_64
5.4.204-113.362.amzn2.x86_64
```

### Kubernetes Version

```
Client Version: v1.24.3
Kustomize Version: v4.5.4
Server Version: v1.22.11-eks-18ef993
```

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

**Comment by mhulscher** [comment]
I want to add that I did not experience the same issue while we were still on 1.11.

**Comment by mhulscher** [comment]
While trying to reproduce this on 1.11.8 I found that cilium also received 401 Unauthorized from removed kube-apiservers. However, it doesn't result in cilium removing identities/endpoints/etc and thus doesn't cause a network outage.

For now we are going to rollback to 1.11.8. I will try and prepare a sysdump of 1.12 while the issue presents itself.

**Comment by m00lecule** [comment]
Hello @mhulscher!

I am probably facing the same issue - does revert to cilium 1.11 helped?

**Comment by mhulscher** [comment]
@m00lecule It did. The problem is not present in 1.11.

**Comment by m00lecule** [comment]
@mhulscher  It's specific patch version or all minor versions (1.11.X)?

**Comment by mhulscher** [comment]
The issue is not present in at least 1.11.{7,8,9}

**Comment by m00lecule** [comment]
FYI: In my case the outage was significant - took down production for few hours. The network connections were hanging until I have manually restarted Cilium pods.

**Comment by jasonaliyetti** [comment]
My team is seeing this too and it caused a substantial outage last week.  We are planning to rollback to 1.11 this week.

**Comment by joestringer** [comment]
Just as a heads up, there are some fixes in a related area coming into Cilium 1.12.3 soon which may impact these changes. Currently without a reliable set of instructions to reproduce, it's a bit hard to investigate or confirm what the issue is. However, I would be interested to hear whether this issue continues to occur on 1.12.3.

**Comment by aditighag** [comment]
>AWS EKS scales its kube-apiserver(s) out and in. I notice that whenever a kube-apiserver is removed, the cilium-agent and -operator continue to make API requests to this now decommissioned kube-apiserver. What's annoying is that this kube-apiserver does not seem to close existing connections. Instead, it now returns a 401 Unauthorized as a response to all requests. This could also be considered an EKS bug.

@mhulscher Thanks for the report. Since you've enabled kube-proxy-replacement, you might be running into an existing limitation of socket-lb. But you should see the same behavior across cilium versions.

So let's try to break down the issue:

(1) Cilium agent continuing to connect to the stale kube-api server (sounds like https://github.com/cilium/cilium/issues/12859). This might explain why restart of the cilium agent fixes the issue as the agent will connect to an active kube-api server. 

(2) Kube-api server returning 401 instead of closing connections seems to be an EKS bug as you pointed out

(3) Why does downgrading to 1.11 Cilium version resolve the issue? Your comment might give a hint - https://github.com/cilium/cilium/issues/20915#issuecomment-1219363906. We'll need to debug this from the angle of identities/policy processing.

To get more clarity around (1), can you please report if the issue exists if you disable kube-proxy replacement?

We'll need more investigation on (3).

**Comment by benesch** [comment]
> Currently without a reliable set of instructions to reproduce, it's a bit hard to investigate or confirm what the issue is. 

@joestringer: this is not hard to reproduce! Turn on an EKS cluster, install Cilium v1.12, and then put sufficient load on the API server that EKS scales it up. The last part is the only part that is somewhat nondeterministic, but I have to imagine that creating a few thousand pods in a tight loop would do the trick every time.

This was described by the original issue poster here:

> I reproduce it myself by creating a new cluster and then scheduling a lot of pods. This often leads to scaling events.

Do you do production testing of Cilium on EKS?

**Comment by mhulscher** [comment]
I just tested with cilium 1.12.3 and I can confirm that the problem is still present.

**Comment by munjalpatel** [comment]
> > Currently without a reliable set of instructions to reproduce, it's a bit hard to investigate or confirm what the issue is.
> 
> @joestringer: this is not hard to reproduce! Turn on an EKS cluster, install Cilium v1.12, and then put sufficient load on the API server that EKS scales it up. The last part is the only part that is somewhat nondeterministic, but I have to imagine that creating a few thousand pods in a tight loop would do the trick every time.
> 
> This was described by the original issue poster here:
> 
> > I reproduce it myself by creating a new cluster and then scheduling a lot of pods. This often leads to scaling events.
> 
> Do you do production testing of Cilium on EKS?

@joestringer I agree with @benesch I just broke this by simply scaling up `huddle-ui` to 20 instances.

**Comment by munjalpatel** [comment]
We will downgrade to v1.11 for now.

**Comment by jose-galvez** [comment]
Our team is also facing this same issue. Is there an ETA on when this might be addressed?

**Comment by gandro** [comment]
There is no ETA, since it's still unclear why Cilium's k8s heartbeat reconnection logic is not triggered.

We've only managed to reproduce it once so far. If anyone has pointers on how to reproduce EKS scale-down very reliably, please let us know. Just scaling the pods (as recommended above) does not always work. It seems there are other parameters taken into account too (node count maybe?).

**Comment by munjalpatel** [comment]
@gandro for us, we had a single node to start with and then Karpenter was spinning up new nodes as needed.

**Comment by mhulscher** [comment]
@gandro perhaps you could consider creating an EKS 1.23 cluster w/ cilium 1.12, then upgrading EKS to 1.24. This will cause the kube-apiservers to be replaced.

**Comment by recollir** [comment]
We are seeing a similar behavior with EKS 1.23 and Cilium 1.10.16. At least the error msg pointed us to this issue. And a restart of the operator solved the problem.

**Comment by recollir** [comment]
@bimmlerd Of course I want to provide some more information. A tl;dr upfront though: it seems that we are not affected by this.

Longer version: We experienced communication problems for workloads as described by @mhulscher in the initial description. This was on EKS 1.23 clusters running 1.10.10 and 1.10.16. The error msgs from the operator corresponded to the error msgs from the initial issue description. We also saw similar issues as mentioned in https://github.com/aws/containers-roadmap/issues/1810. Hence me adding myself to this thread before being able to look closer into this. I just wanted to loose track of this here.

As it now turns out, after looking more into this yesterday, we see that we get this problems only while enabling/associating an oidc identity provider for the cluster. After the enabling process has finished the error msgs disappear and network connectivity is restored. At least from some few initial tests yesterday.

Sorry for causing confusion.

**Comment by gandro** [comment]
We did some more testing and we're still not able to reproduce the behavior where "cilium [removes] identities/endpoints/etc and thus doesn't cause a network outage" as mentioned in the original report.

Upgrading EKS K8s version indeed seems like reliable method to swap out the apiserver endpoint (thanks for that tip @mhulscher), it doesn't cause any network outages in our testing. `cilium connectivity test` runs just fine.

Yes, the operator (and agent) temporarily looses connection to the apiserver, as evident by the `Unauthorized` messages - but they eventually (1-10 minutes in our testing) reconnect once the TCP connection to the old endpoints is interrupted. With PR #22393, that down-time should be significantly reduced in the future (since it will force a reconnect much earlier).

Overall however, we have not been able to observe any outages in terms of connectivity. So it would be great if we could get some more details as to what specifically is breaking during scale-down, so we can investigate those further (and understand why it would behave differently on 1.11 vs 1.12).

**Comment by mhulscher** [comment]
@gandro during your testing, did you install cilium w/ a similar configuration to what I shared in the first post?

Something to keep in mind: when an EKS upgrade is finished, the old kube-apiserver endpoints are not necessarily immediately removed. As such, the issue might not present itself. The issue occurs when the old kube-apiservers are removed, which happens several minutes after the upgrade is completed. This is why I keep a watch on `kubectl -n default get ep kubernetes` to keep see when apiservers are removed.

**Comment by gandro** [comment]
> @gandro during your testing, did you install cilium w/ a similar configuration to what I shared in the first post?

Yes, we just did another attempt with basically exactly your config on Cilium v1.12.3 and were not able to reproduce any connectivity issues (using the `examples/connectivity-check` yaml). 

While the K8s endpoints where replaced (triggered by an K8s upgrade in our case), all pods were able to communicate just fine while the operator reconnected.

Therefore, we need more information as to how exactly the network outage looks like:
 - What pods are affected?
 - Are they talking to endpoints on the removed nodes? To new nodes?
 - Do the affected pods have policies applied?
 - What kind of endpoints are affected? pod to pod, pod to service, pod to host, etc

> Something to keep in mind: when an EKS upgrade is finished, the old kube-apiserver endpoints are not necessarily immediately removed. As such, the issue might not present itself. The issue occurs when the old kube-apiservers are removed, which happens several minutes after the upgrade is completed. This is why I keep a watch on `kubectl -n default get ep kubernetes` to keep see when apiservers are removed.

Yes, I can confirm that we observe this as well. But as mentioned before, in our testing, the operator (and the agents) eventually reconnect once the old endpoints are fully removed. With PR https://github.com/cilium/cilium/pull/22393 they will reconnect even earlier. But this behavior is present on v1.11 and v1.12, so it cannot be the cause for the issues reported here.

**Comment by mhulscher** [comment]
It's a pity you can't reproduce it reliably. Unfortunately, I no longer work on the systems where I first encountered this issue, so I can't reproduce it myself for some time.

> What pods are affected?

Can't say for sure. What I can say, is that the cilium-agent pods themselves start failing their readiness probes when the issue occurs.

> Are they talking to endpoints on the removed nodes? To new nodes?

I am not sure what you mean. There are no removed or added worker nodes.

> Do the affected pods have policies applied?

Yes. All pods in the cluster are subject to CCNPs. Also, all pods are subject to a CiliumLocalRedirectPolicy for NodeLocalDNS. Furthermore, all DNS requests pass through the DNS proxy for layer 7 inspection.

> What kind of endpoints are affected? pod to pod, pod to service, pod to host, etc

I can only say for sure that traffic toward destinations outside the cluster is affected.

**Comment by eugenestarchenko** [comment]
We're seeing this issue as well with operator-generic:v1.12.3-cee.1 & cilium:v1.12.3-cee.1 
It causes a domino effect.
Any suggestions from the Cilium team? 
Looks like the proposed fix that was backported(merged) is not included in 1.12.4 release
https://github.com/cilium/cilium/blob/1.12.4/pkg/k8s/watchers/watcher.go#L296-L300

**Comment by gandro** [comment]
> We're seeing this issue as well with operator-generic:v1.12.3-cee.1 & cilium:v1.12.3-cee.1 It causes a domino effect. Any suggestions from the Cilium team? 

What issues specifically do you see? There are multiple factors:

- The unnecessarily long control plane unavailability - this will be fixed in v1.12.5. As a potential workaround, one could decrease the K8s heartbeat duration, e.g. `extraConfig.k8s-heartbeat-timeout=5s` - this will however increase traffic to the k8s apiserver.
- If you do observe network outages, please provide details (ideally even a sysdump), since those are likely not addressed by the change above. What kind of workloads are affected? Do you use a particular DNS setup (our current working hypothesis involves DNS somehow also being unavailable too during the k8s-apiserver endpoint migration)? What are the domino effects that you mentioned? Thanks!

Please note: Even with the fixes and workarounds, you will still see the `Unauthorized` log line. The error message itself however is not necessarily a problem, since Cilium should reconnect. Only if the log line coincides with other problems such as operator being unavailable or network outages does it indicate that something else is failing as well.

**Comment by arusahni** [comment]
I was able to reproduce this issue with Cilium 1.12.1, but not Cilium 1.12.5

My methodology:

1. Spin up a fresh cluster running my org's prod workload with Cilium 1.11.11 and k8s 1.22.
2. Create an environment.
3. Verify everything is connectable and that the cluster looks healthy.
4. Upgrade the stack to Cilium 1.12.x.
5. Verify everything is connectable and the cluster looks healthy.
6. Upgrade to k8s 1.23

After the k8s upgrade I was able to see  I see my org's pods stuck in a ContainerCreating state with events such as:

> failed to set up sandbox container "8803d905ce25a431631bb389fcb61892f5e6cfe4ef233478c9ff3b189e823950" network for pod "[pod_name]-jfpj4": networkPlugin cni failed to set up pod "[pod_name]-jfpj4_system" network: Unable to create endpoint: Cilium API client timeout exceeded

and

>    Warning  FailedCreatePodSandBox  80s (x364 over 102m)  kubelet  (combined from similar events): Failed to create pod sandbox: rpc error: code = Unknown desc = failed to set up sandbox container "4b55c0425f6947b27d06a75afb0b1db710445
 e232e5a82fb24db1fe9760ed5d7" network for pod "[pod_name]-5wngq": networkPlugin cni failed to set up pod "[pod_name]-5wngq_system" network: Unable to create endpoint: [PUT /endpoint/{id}][429] putEndpointIdTooManyRequests                                                                                                                                                                                                       

Additionally, the node's Cilium agent was logging errors saying it couldn't fetch the labels for the container stuck in that state. Restarting the agent and operator on the node w/the stuck pod did not unjam things. I had to restart all of the operators and agents in the cluster in order to restore connectivity.

**Comment by gandro** [comment]
@arusahni Thanks a lot for providing the methodology. A few questions on this:

- Are you able to provide a full sysdump for that test setup? The snippets provided give us a good overview over the symptoms, but it's hard to establish what caused them. Feel free to reach out to me privately to share the sysdump, assuming you cannot upload it publicly here
- When you say you cannot reproduce this with 1.12.5, does that mean that the k8s upgrade goes though smoothly on 1.12.5 but not on 1.12.1?
- What do you mean with step 2 "Create an environment."?

Thanks! We'll try to reproduce following your steps, but they look very similar to what we've already (unsuccessfully) tried as potential reproduction setups, so there are likely more variables to this.

**Comment by mhulscher** [comment]
Using the latest cilium 1.12.6 the problem seems to persist. Because all DNS traffic goes thru the l7-proxy for dns inspection, it looks like all DNS in the cluster are failing. This is from an affected cilium-agent:

```
cilium-2k2kp cilium-agent level=info msg="Waiting for endpoint to be generated" containerID= datapathPolicyRevision=0 desiredPolicyRevision=0 endpointID=1861 identity=54592 ipv4= ipv6= k8sPodName=/ subsys=endpoint
cilium-2k2kp cilium-agent level=info msg="Rewrote endpoint BPF program" containerID= datapathPolicyRevision=0 desiredPolicyRevision=11 endpointID=1861 identity=54592 ipv4= ipv6= k8sPodName=/ subsys=endpoint
cilium-2k2kp cilium-agent level=info msg="Successful endpoint creation" containerID= datapathPolicyRevision=11 desiredPolicyRevision=11 endpointID=1861 identity=54592 ipv4= ipv6= k8sPodName=/ subsys=daemon
cilium-2k2kp cilium-agent level=info msg="regenerating all endpoints" reason="one or more identities created or deleted" subsys=endpoint-manager
cilium-2k2kp cilium-agent level=info msg="regenerating all endpoints" reason="one or more identities created or deleted" subsys=endpoint-manager
cilium-2k2kp cilium-agent level=info msg="Local Redirect Policy updates are not handled" subsys=k8s-watcher
cilium-2k2kp cilium-agent level=info msg="regenerating all endpoints" reason="one or more identities created or deleted" subsys=endpoint-manager
cilium-2k2kp cilium-agent level=info msg="regenerating all endpoints" reason="one or more identities created or deleted" subsys=endpoint-manager
cilium-2k2kp cilium-agent level=warning msg="No response from probe within 15 seconds" probe=l7-proxy subsys=status
cilium-2k2kp cilium-agent level=warning msg="Timeout while waiting probe" probe=l7-proxy startTime="2023-01-31 11:08:54.200027098 +0000 UTC m=+2021.262533826" subsys=status
```

This is from a pod (rancher) that is attempting to talk to a resource on the internet whose traffic is allowed:


```
rancher-54c4f6b49f-7c7xk rancher 2023/01/31 11:10:48 [ERROR] error syncing 'library': handler catalog: Remote commit check failed: Repo [https://git.rancher.io/charts] is not accessible: Get "https://git.rancher.io/repos/charts/commits/master": context deadline exceeded (Client.Timeout exceeded while awaiting headers), requeuing
rancher-54c4f6b49f-7c7xk rancher 2023/01/31 11:10:48 [ERROR] error syncing 'helm3-library': handler catalog: Remote commit check failed: Repo [https://git.rancher.io/helm3-charts] is not accessible: Get "https://git.rancher.io/repos/helm3-charts/commits/master": context deadline exceeded (Client.Timeout exceeded while awaiting headers), requeuing
rancher-54c4f6b49f-7c7xk rancher 2023/01/31 11:10:48 [ERROR] error syncing 'system-library': handler catalog: Remote commit check failed: Repo [https://git.rancher.io/system-charts] is not accessible: Get "https://git.rancher.io/repos/system-charts/commits/release-v2.7": context deadline exceeded (Client.Timeout exceeded while awaiting headers), requeuing
rancher-54c4f6b49f-7c7xk rancher 2023/01/31 11:10:58 [ERROR] error syncing 'system-library': handler catalog: Remote commit check failed: Repo [https://git.rancher.io/system-charts] is not accessible: Get "https://git.rancher.io/repos/system-charts/commits/release-v2.7": dial tcp: lookup git.rancher.io on 172.20.0.10:53: read udp 10.3.231.12:57197->10.3.12.173:53: i/o timeout, requeuing
rancher-54c4f6b49f-7c7xk rancher 2023/01/31 11:10:58 [ERROR] error syncing 'library': handler catalog: Remote commit check failed: Repo [https://git.rancher.io/charts] is not accessible: Get "https://git.rancher.io/repos/charts/commits/master": dial tcp: lookup git.rancher.io on 172.20.0.10:53: read udp 10.3.231.12:57197->10.3.12.173:53: i/o timeout, requeuing
rancher-54c4f6b49f-7c7xk rancher 2023/01/31 11:10:58 [ERROR] error syncing 'helm3-library': handler catalog: Remote commit check failed: Repo [https://git.rancher.io/helm3-charts] is not accessible: Get "https://git.rancher.io/repos/helm3-charts/commits/master": dial tcp: lookup git.rancher.io on 172.20.0.10:53: read udp 10.3.231.12:57197->10.3.12.173:53: i/o timeout, requeuing
```

**Comment by gandro** [comment]
@mhulscher Thanks for the update!

```
2023-01-31T11:07:46.392482213Z level=warning msg="Timeout while waiting probe" probe=l7-proxy startTime="2023-01-31 11:06:46.391301457 +0000 UTC m=+1901.287024906" subsys=status
```

Indicates an issue unrelated to the EKS K8s apiserver, since that l7proxy codepath does not involve the K8s apiserver. Instead, this likely is due to a deadlock we recently discovered: https://github.com/cilium/cilium/pull/23377

The fix for that one should be available in v1.12.7 scheduled for mid-February.

**Comment by mhulscher** [comment]
@gandro Could the hanging of the l7-proxy also explain why the cilium-agent's readiness/liveness probe start failing? Because that's what's happening. Somehow this is triggered only after a kube-apiserver endpoint is removed.

```
$ k logs --timestamps -p cilium-2k2kp

...

023-01-31T11:07:15.221789344Z level=info msg="Local Redirect Policy updates are not handled" subsys=k8s-watcher
2023-01-31T11:07:31.184611613Z level=info msg="regenerating all endpoints" reason="one or more identities created or deleted" subsys=endpoint-manager
2023-01-31T11:08:51.990922365Z level=info msg="regenerating all endpoints" reason="one or more identities created or deleted" subsys=endpoint-manager
2023-01-31T11:09:09.201081428Z level=warning msg="No response from probe within 15 seconds" probe=l7-proxy subsys=status
2023-01-31T11:09:54.201032864Z level=warning msg="Timeout while waiting probe" probe=l7-proxy startTime="2023-01-31 11:08:54.200027098 +0000 UTC m=+2021.262533826" subsys=status
2023-01-31T11:10:05.147696700Z level=warning msg="hubble events queue is full: dropping messages; consider increasing the queue size (hubble-event-queue-size) or provisioning more CPU" subsys=hubble
2023-01-31T11:14:51.764283547Z level=info msg="Exiting due to signal" signal=terminated subsys=daemon
2023-01-31T11:14:51.764907655Z level=info msg="Waiting for all endpoints' go routines to be stopped." subsys=daemon
2023-01-31T11:14:51.764955351Z level=info msg="Shutting down... " subsys=daemon
2023-01-31T11:14:51.764961431Z level=info msg="HTTP server Shutdown: context deadline exceeded" subsys=daemon
2023-01-31T11:14:51.764964573Z level=info msg="Stopped serving cilium API at unix:///var/run/cilium/cilium.sock" subsys=daemon
2023-01-31T11:14:51.764967158Z level=info msg="Stopping fswatcher" config=tls-server subsys=hubble
2023-01-31T11:14:51.764969677Z level=info msg="Stopped reading perf buffer" startTime="2023-01-31 10:35:24.015406915 +0000 UTC m=+11.077913635" subsys=monitor-agent
```

```
$ kubectl events --for pod/cilium-2k2kp --types=Warning -o json | jq '.items[] | {message: .message, firstTimestamp: .firstTimestamp}'

{
  "message": "Liveness probe failed: HTTP probe failed with statuscode: 503",
  "firstTimestamp": "2023-01-31T11:10:21Z"
}
{
  "message": "Readiness probe failed: HTTP probe failed with statuscode: 503",
  "firstTimestamp": "2023-01-31T11:10:21Z"
}
```

And here is the moment the last non-current kube-apiserver was present:

![image](https://user-images.githubusercontent.com/18721288/215750336-5350f17d-6442-4544-865b-653d00ce8a38.png)

Which corresponds more or less with the time of the l7-proxy failure, given that my prometeus scrape interval is 30s.

```
$ TZ=UTC date -d @1675163340
di 31 jan 2023 11:09:00 UTC
```

**Comment by gandro** [comment]
> @gandro Could the hanging of the l7-proxy also explain why the cilium-agent's readiness/liveness probe start failing? 

Yes. This is how we discovered the deadlock in the first place. The only place where the l7proxy probe should block is while taking the lock in `GetStatusModel()`. Thus, if that probe times out, it's very likely due to a deadlock.
https://github.com/cilium/cilium/blob/a1442e0d6d0601fa09a59b7f322b2aead4edc37e/pkg/proxy/proxy.go#L718-L720

The l7proxy probe timeout then causes "cilium status" to report stale probes:
https://github.com/cilium/cilium/blob/314ca7baeff4f568ffc0bad95124e40665b1f88c/pkg/status/status.go#L232

Which marks the pod as unhealthy in the liveness probe:
https://github.com/cilium/cilium/blob/5c072af7af28a0ad14de67e5b7506bb7fef90352/daemon/cmd/status.go#L652-L657

> Because that's what's happening. Somehow this is triggered only after a kube-apiserver endpoint is removed.

This is an interesting correlation. Unfortunately, due to the deadlock we don't see which l7proxy operation failed and caused the deadlock, but it seems that the l7proxy operation might be indeed triggered by the k8s apiserver being removed.

With the https://github.com/cilium/cilium/pull/23377 patch applied, we should get more info. Alternatively, a sysdump in that 5 minute period where the livenesscheck is failing (but before the pod is restarted) would also be super helpful do debug this (since again, we haven't been able to reproduce this ourselves so far).

**Comment by mhulscher** [comment]
I am eagerly awaiting the new patch release. If I encounter the problem again I'll try and make a sysdump. I'll reach out on slack to see if I can share it privately.

**Comment by gandro** [comment]
Thanks to @mhulscher providing us with a sysdump, we (@bimmlerd and I) finally have some understanding as to what seems to be happening here. Unfortunately, the fix provided in #23377 is not enough to avoid the issue.

TL;DR: There seems to be a deadlock between IPCache and some endpoint lock. This seems to occur if there are L7 policies in place, some endpoint regeneration happens at an inopportune time, and the kube-apiserver endpoints change (which is why this can be triggered by an EKS scale-down event). Because multiple subsystems are affected by the deadlock, there are wide range of symptoms (proxied DNS lookups might fail, endpoint regeneration is stuck, and eventually the l7proxy will cause the DaemonSet's liveness probe to fail after 5 minutes, causing an agent restart. After the agent restart, the symptoms should go away).

Looking at the gops threaddump, there are two goroutines which blocked on each other: The K8s endpoint watcher and an endpoint regeneration goroutine for the endpoint `0x4000f4b880`.

The deadlock happens as follows:

 1. The K8s endpoint watcher goroutine observes a change in the `kubernetes` endpoints. This results in it updating IPCache IP Identity Metadata labels for any IPs with the `kube-apiserver` label (`handleKubeAPIServerServiceEPChanges`). This calls into IPCache via `IPCache.removeLabelsFromIPs`, which in *holds on to the IPCache mutex*. It then spawns a new child go routine in `EndpointManager.UpdatePolicyMaps`. Notably, IPCache waits on this `UpdatePolicyMaps` child go routine to return, while holding on to the IPCache mutex. The `UpdatePolicyMaps` goroutine iterates over all endpoints, attempting to lock them. Unfortunately, one of the endpoints in our stackdump, namely the one at addr `0x4000f4b880` cannot be locked, since the locks seems in use by another go routine.
 
 2. That other go routine is the endpoint regeneration goroutine for ep `0x4000f4b880`. We don't know why regeneration was triggered (potentially due to an unrelated endpoint deletion?) yet, but we see that regeneration is in progress. The regeneration calls `Endpoint.runPreCompilationSteps` which *holds on to the Endpoint 0x4000f4b880 mutex*. It eventually calls into `Proxy.CreateOrUpdateRedirect`, holds on to the proxy lock (which is why the l7probe eventually fails), and calls into `DNSProxy.GetRules`. The `DNSProxy.GetRules` function then needs to obtain the identities of certain IPs, and it does this by calling into `IPCache.LookupByIdentity`. This is where the deadlock happens. `IPCache.LookupByIdentity` attempts to take the IPCache mutex. But that one is already taken by the "K8s endpoint watcher" goroutine, which is waiting on us to release the lock for endpoint 0x4000f4b880. A classical deadlock.
 

Unfortunately, we don't know yet how to fix this. The newer code here is the `IPCache.removeLabelsFromIPs`, so it might be possible to fix it there somehow. We'll continue to work on a solution.



<details><summary>Stacktrace for the first goroutine ("the K8s endpoint watcher"):</summary>

```
goroutine 24422 [semacquire, 2 minutes]:
sync.runtime_SemacquireMutex(0x4000cd2d08?, 0xc0?, 0x4000cd2d28?)
	/usr/local/go/src/runtime/sema.go:71 +0x28
sync.(*Mutex).lockSlow(0x4000f4b8d0)
	/usr/local/go/src/sync/mutex.go:162 +0x180
sync.(*Mutex).Lock(...)
	/usr/local/go/src/sync/mutex.go:81
sync.(*RWMutex).Lock(0x4000f4b8d0?)
	/usr/local/go/src/sync/rwmutex.go:139 +0x7c
github.com/cilium/cilium/pkg/endpoint.(*Endpoint).lockAlive(0x4000f4b880)
	/go/src/github.com/cilium/cilium/pkg/endpoint/lock.go:14 +0x34
github.com/cilium/cilium/pkg/endpoint.(*Endpoint).ApplyPolicyMapChanges(0x4000f4b880, 0x77d0c?)
	/go/src/github.com/cilium/cilium/pkg/endpoint/bpf.go:1165 +0x34
github.com/cilium/cilium/pkg/endpointmanager.(*EndpointManager).UpdatePolicyMaps.func2(0x4000965500?)
	/go/src/github.com/cilium/cilium/pkg/endpointmanager/manager.go:171 +0x50
created by github.com/cilium/cilium/pkg/endpointmanager.(*EndpointManager).UpdatePolicyMaps
	/go/src/github.com/cilium/cilium/pkg/endpointmanager/manager.go:168 +0x158
	
	|| vv connected via go routine spawn ^^ ||

goroutine 249 [semacquire, 2 minutes]:
sync.runtime_Semacquire(0x4002c56130?)
	/usr/local/go/src/runtime/sema.go:56 +0x2c
sync.(*WaitGroup).Wait(0x4002c56130)
	/usr/local/go/src/sync/waitgroup.go:136 +0x88
github.com/cilium/cilium/pkg/ipcache.(*IPCache).UpdatePolicyMaps(0x40003a3e80, {0x2e9a3c0, 0x400018e008}, 0xb?, 0x40040c3170)
	/go/src/github.com/cilium/cilium/pkg/ipcache/metadata.go:235 +0xcc
github.com/cilium/cilium/pkg/ipcache.(*IPCache).removeLabelsFromIPs(0x40003a3e80, 0x4001c87758?, {0x295c64b, 0xf})
	/go/src/github.com/cilium/cilium/pkg/ipcache/metadata.go:414 +0x618
github.com/cilium/cilium/pkg/ipcache.(*IPCache).RemoveLabelsExcluded(0x40003a3e80, 0x40004a1c50, 0x4001c878b8?, {0x295c64b, 0xf})
	/go/src/github.com/cilium/cilium/pkg/ipcache/metadata.go:328 +0x194
github.com/cilium/cilium/pkg/k8s/watchers.(*K8sWatcher).handleKubeAPIServerServiceEPChanges(0x40001a7400, 0x4001c87988?)
	/go/src/github.com/cilium/cilium/pkg/k8s/watchers/endpoint.go:135 +0x50
github.com/cilium/cilium/pkg/k8s/watchers.(*K8sWatcher).addKubeAPIServerServiceEPSliceV1(0x4000b76300?, 0x40050b10a0)
	/go/src/github.com/cilium/cilium/pkg/k8s/watchers/endpoint_slice.go:205 +0x3c8
github.com/cilium/cilium/pkg/k8s/watchers.(*K8sWatcher).updateK8sEndpointSliceV1(0x40001a7400, 0x40050b10a0?, 0x40050b10a0?)
	/go/src/github.com/cilium/cilium/pkg/k8s/watchers/endpoint_slice.go:178 +0x6c
github.com/cilium/cilium/pkg/k8s/watchers.(*K8sWatcher).endpointSlicesInit.func2({0x28ee5c0?, 0x4000d43960?}, {0x28ee5c0, 0x40050b10a0})
	/go/src/github.com/cilium/cilium/pkg/k8s/watchers/endpoint_slice.go:71 +0xd8
k8s.io/client-go/tools/cache.ResourceEventHandlerFuncs.OnUpdate(...)
	/go/src/github.com/cilium/cilium/vendor/k8s.io/client-go/tools/cache/controller.go:239
github.com/cilium/cilium/pkg/k8s/informer.NewInformerWithStore.func1({0x2475e40?, 0x4000a51050?})
	/go/src/github.com/cilium/cilium/pkg/k8s/informer/informer.go:118 +0x2b0
k8s.io/client-go/tools/cache.(*DeltaFIFO).Pop(0x4000b265a0, 0x4000354700)
	/go/src/github.com/cilium/cilium/vendor/k8s.io/client-go/tools/cache/delta_fifo.go:554 +0x50c
k8s.io/client-go/tools/cache.(*controller).processLoop(0x40004e3680)
	/go/src/github.com/cilium/cilium/vendor/k8s.io/client-go/tools/cache/controller.go:184 +0x40
k8s.io/apimachinery/pkg/util/wait.BackoffUntil.func1(0x45b8?)
	/go/src/github.com/cilium/cilium/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:155 +0x44
k8s.io/apimachinery/pkg/util/wait.BackoffUntil(0x1?, {0x2e6fc60, 0x4000ad4630}, 0x1, 0x400012fd40)
	/go/src/github.com/cilium/cilium/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:156 +0x94
k8s.io/apimachinery/pkg/util/wait.JitterUntil(0x40004e36e8?, 0x3b9aca00, 0x0, 0xa0?, 0xffff89776e68?)
	/go/src/github.com/cilium/cilium/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:133 +0x84
k8s.io/apimachinery/pkg/util/wait.Until(...)
	/go/src/github.com/cilium/cilium/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:90
k8s.io/client-go/tools/cache.(*controller).Run(0x40004e3680, 0x400012fd40)
	/go/src/github.com/cilium/cilium/vendor/k8s.io/client-go/tools/cache/controller.go:155 +0x28c
created by github.com/cilium/cilium/pkg/k8s/watchers.(*K8sWatcher).endpointSlicesInit
	/go/src/github.com/cilium/cilium/pkg/k8s/watchers/endpoint_slice.go:156 +0x63c
```

</details>



<details><summary>Stacktrace for the second go routine ("the ep 0x4000f4b880 regeneration go routine").:</summary>

```
goroutine 2529 [select, 2 minutes]:
golang.org/x/sync/semaphore.(*Weighted).Acquire(0x4000476190, {0x2e9a3c0, 0x400018e000}, 0x1)
	/go/src/github.com/cilium/cilium/vendor/golang.org/x/sync/semaphore/semaphore.go:60 +0x410
github.com/cilium/cilium/pkg/lock.(*SemaphoredMutex).RLock(...)
	/go/src/github.com/cilium/cilium/pkg/lock/semaphored_mutex.go:46
github.com/cilium/cilium/pkg/ipcache.(*IPCache).LookupByIdentity(0x40003a3e80, 0x1fcf910?)
	/go/src/github.com/cilium/cilium/pkg/ipcache/ipcache.go:639 +0x64
github.com/cilium/cilium/daemon/cmd.(*Daemon).lookupIPsBySecID(0x400133f7a8?, 0x1fcf7f0?)
	/go/src/github.com/cilium/cilium/daemon/cmd/fqdn.go:404 +0x24
github.com/cilium/cilium/pkg/fqdn/dnsproxy.(*DNSProxy).GetRules(0x4000779a40, 0x723)
	/go/src/github.com/cilium/cilium/pkg/fqdn/dnsproxy/proxy.go:200 +0x35c
github.com/cilium/cilium/pkg/proxy.(*dnsRedirect).setRules(0x4000125400, 0x0?, 0x400334af30)
	/go/src/github.com/cilium/cilium/pkg/proxy/dns.go:43 +0x1c4
github.com/cilium/cilium/pkg/proxy.(*dnsRedirect).UpdateRules(0x4000125400, 0x0?)
	/go/src/github.com/cilium/cilium/pkg/proxy/dns.go:58 +0x34
github.com/cilium/cilium/pkg/proxy.(*Proxy).CreateOrUpdateRedirect(0x40001a1d40, {0x2e9a388?, 0x4001549800}, {0x2e9b380, 0x40047fb580}, {0x4003897890, 0x12}, {0x2eaf9c8, 0x4000f4b880}, 0x40036e93c0)
	/go/src/github.com/cilium/cilium/pkg/proxy/proxy.go:504 +0x674
github.com/cilium/cilium/pkg/endpoint.(*Endpoint).addNewRedirectsFromDesiredPolicy(0x4000f4b880, 0x0, 0x2935fbb?, 0x6?)
	/go/src/github.com/cilium/cilium/pkg/endpoint/bpf.go:253 +0x238
github.com/cilium/cilium/pkg/endpoint.(*Endpoint).addNewRedirects(0x4000f4b880, 0x1ca98?)
	/go/src/github.com/cilium/cilium/pkg/endpoint/bpf.go:458 +0x18c
github.com/cilium/cilium/pkg/endpoint.(*Endpoint).runPreCompilationSteps(0x4000f4b880, 0x4002ec4c00)
	/go/src/github.com/cilium/cilium/pkg/endpoint/bpf.go:828 +0x384
github.com/cilium/cilium/pkg/endpoint.(*Endpoint).regenerateBPF(0x4000f4b880, 0x4002ec4c00)
	/go/src/github.com/cilium/cilium/pkg/endpoint/bpf.go:579 +0x14c
github.com/cilium/cilium/pkg/endpoint.(*Endpoint).regenerate(0x4000f4b880, 0x4002ec4c00)
	/go/src/github.com/cilium/cilium/pkg/endpoint/policy.go:398 +0x5fc
github.com/cilium/cilium/pkg/endpoint.(*EndpointRegenerationEvent).Handle(0x4002ca7580, 0x15548?)
	/go/src/github.com/cilium/cilium/pkg/endpoint/events.go:53 +0x2b0
github.com/cilium/cilium/pkg/eventqueue.(*EventQueue).run.func1()
	/go/src/github.com/cilium/cilium/pkg/eventqueue/eventqueue.go:245 +0x10c
sync.(*Once).doSlow(0x2ea3e98?, 0x4000f09500?)
	/usr/local/go/src/sync/once.go:68 +0x10c
sync.(*Once).Do(...)
	/usr/local/go/src/sync/once.go:59
github.com/cilium/cilium/pkg/eventqueue.(*EventQueue).run(0x2ea3e98?)
	/go/src/github.com/cilium/cilium/pkg/eventqueue/eventqueue.go:233 +0x4c
created by github.com/cilium/cilium/pkg/eventqueue.(*EventQueue).Run
	/go/src/github.com/cilium/cilium/pkg/eventqueue/eventqueue.go:229 +0x80
```

</details>

**Comment by gandro** [comment]
The deadlock has unfortunately not been fully fixed in v1.12.8 - there is a follow-up PR needed, which hopefully will make it into the next release: https://github.com/cilium/cilium/pull/24672
